README.md
# cargo-compiler-interrupts

[![crates.io](https://img.shields.io/crates/v/cargo-compiler-interrupts.svg)][crates.io]
[![docs.rs](https://docs.rs/cargo-compiler-interrupts/badge.svg)][docs.rs]
[![license](https://img.shields.io/crates/l/cargo-compiler-interrupts.svg)][license]

`cargo-compiler-interrupts` provides you a seamless way to integrate the
[Compiler Interrupts][compiler-interrupts-paper] to any Rust packages.
Check out the Compiler Interrupts [main repository][compiler-interrupts] for more info.

## Requirements

* [Rust 1.45.0][rust] or later and [LLVM 9][llvm] or later are required.
Both must have the same LLVM major version.
* You can check the LLVM version from Rust toolchain and LLVM toolchain by running `rustc -vV`
and `llvm-config --version` respectively.
* x86-64 architecture with Linux or macOS is highly recommended.
Other architectures and platforms have not been tested.

## Installation

`cargo-compiler-interrupts` can be installed via `cargo install`.

``` sh
cargo install cargo-compiler-interrupts
```

You can also fetch the repo and install  using `--path`.

``` sh
git clone https://github.com/bitslab/cargo-compiler-interrupts
cargo install --path ./cargo-compiler-interrupts
```

Install the Compiler Interrupts library.

``` sh
cargo-lib-ci install
```

## Getting started

### Commands

* `cargo-lib-ci` — manage the Compiler Interrupts library.
* `cargo-build-ci` — build and integrate the Compiler Interrupts to the package.
* `cargo-run-ci` — run the integrated binary.
You can specify which binary to run by passing `--bin <BINARY>`.

``` sh
cargo-lib-ci install    # install the CI library
cargo-build-ci          # build and integrate CI to the binary
cargo-run-ci            # run the CI-integrated binary
```

### Integration

* Run `cargo-lib-ci install` to install the Compiler Interrupts library first.
* Before running `cargo-build-ci`, add the Compiler Interrupts API package as the dependency for
your Rust program and registers the Compiler Interrupts handler.
Compiler Interrupts API is provided through the [`compiler-interrupts`][compiler-interrupts-rs]
crate.

``` rust
fn interrupt_handler(ic: i64) {
    println!("Compiler interrupt called with instruction count: {}", ic);
}

unsafe {
    compiler_interrupts::register(1000, 1000, interrupt_handler);
}
```

For more detailed usages and internals, run the command with `--help` option and
check out the **[documentation]**.

## Contribution

All issue reports, feature requests, pull requests and GitHub stars are welcomed
and much appreciated. Issues relating to the Compiler Interrupts library
should be reported to the [main repository][compiler-interrupts].

## Author

Quan Tran ([@quanshousio][quanshousio])

## Acknowledgements

* My advisor [Jakob Eriksson][jakob] for the enormous support for this project.
* [Nilanjana Basu][nilanjana] for implementing the Compiler Interrupts.

## License

`cargo-compiler-interrupts` is available under the MIT license.
See the [LICENSE][license] file for more info.

[crates.io]: https://crates.io/crates/cargo-compiler-interrupts
[docs.rs]: https://docs.rs/cargo-compiler-interrupts
[license]: https://github.com/bitslab/cargo-compiler-interrupts/blob/main/LICENSE
[documentation]: https://github.com/bitslab/cargo-compiler-interrupts/blob/main/DOCUMENTATION.md
[compiler-interrupts]: https://github.com/bitslab/CompilerInterrupts
[compiler-interrupts-rs]: https://github.com/bitslab/compiler-interrupts-rs
[compiler-interrupts-paper]: https://dl.acm.org/doi/10.1145/3453483.3454107
[rust]: https://www.rust-lang.org/tools/install
[llvm]: https://releases.llvm.org
[quanshousio]: https://quanshousio.com
[jakob]: https://www.linkedin.com/in/erikssonjakob
[nilanjana]: https://www.linkedin.com/in/nilanjana-basu-99027959


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. `cargo-compiler-interrupts` adheres to [Semantic Versioning](https://semver.org/).

#### 3.x Releases

- `4.0.x` Releases - [4.0.0](#400) | [4.0.1](#401)
- `3.2.x` Releases - [3.2.0](#320) | [3.2.1](#321)
- `3.1.x` Releases - [3.1.0](#310)
- `3.0.x` Releases - [3.0.0](#300) | [3.0.1](#301)

#### 2.x Releases

- `2.1.x` Releases - [2.1.0](#210)
- `2.0.x` Releases - [2.0.0](#200)

#### 1.x Releases

- `1.0.x` Releases - [1.0.0](#100)

---

## [4.0.1](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/4.0.0)

Released on 2022-10-19.

#### Added

- Add `Cargo.lock` to lock dependency version.

## [4.0.0](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/4.0.0)

#### Added

- Major code refactoring:
  - New subcommands and options. Please check `--help` for further usage.
  - New `cargo` module contains `cargo`-related functions.
  - New `paths` module contains functions for manipulating `Path` and `PathBuf`.
  - New `llvm` module is reimplemented from `util:toolchain`.
  - New `args` module updated with clap 4.0, replaced the old `opts` module.
  - Better logging and error messages.

#### Updated

- Use Rust 2021.
- Major dependency updates such as clap 4.0, indicatif 0.17, and other dependencies.

#### Removed

- Local `cargo-util` package is replaced by using the one from upstream.

---

## [3.2.1](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/3.2.1)

Released on 2021-08-04.

#### Updated

- Documentation and dependencies.
- `cargo-build-ci`
  - Progress indicator is more aware of terminal size changes.

## [3.2.0](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/3.2.0)

Released on 2021-07-31.

#### Added

- `cargo-build-ci`
  - Support integration for `examples` target by passing `--example` option.
- `cargo-run-ci`
  - Ability to pass arguments to the binary by passing `--args` option.

#### Updated

- Various path-related helper functions are renamed.
- `cargo-build-ci`
  - Fix a race condition between integration threads and progress indicator thread when integration thread failed unexpectedly.
  - Failed integration can now be run again without using `cargo clean`.
- `cargo-lib-ci`
  - Fix wrong output formatting while the progress indicator is showing.

#### Removed

- `deps-ci` folder has been removed. All CI-integrated artifacts are now resided in the same directory as the original one.

## [3.1.0](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/3.1.0)

Released on 2021-07-24.

#### Added

- Compiler Interrupts library officially supports LLVM from 9 to 12.
- Documentation for all public items.
- Configuration has three new fields: Path to the debug-enabled library, checksum of the source and remote URL for the source code.
- `cargo-build-ci`
  - Ability to skip crates while integration by passing `--skip-crates` option.
  - Ability to enable debugging mode when integrating by passing `--debug-ci` option.
  - Real-time output from internal `cargo build` invocation.
  - Progress indicator when running the integration.
  - Linker invocations now run concurrently.
  - Output from internal commands such as `opt` and linker is now truncated by default to prevent polluting the terminal. Passing `--debug-ci` enables full and verbose logging.
  - When running with `--debug-ci` option, output from failed process is written to a log file instead of directly printing to the terminal. Log files can be found in the same directory with the configuration file.
- `cargo-lib-ci`
  - Ability to update the library by passing `--update` option.
  - Ability to specify a remote URL to the source code of the library by passing `--url` option.
  - Installing the library now installs both with and without debugging mode versions.
  - File name for the library now has a checksum appended to it for versioning.
  - Progress indicator when installing/updating the library.

#### Updated

- `args` module is renamed to `opts`, same for `Args` structs.
- Using an unsupported version of LLVM will now throw an appropriate error.
- Various helper functions are renamed.
- `cargo-build-ci`
  - Fix the mismatch relocation symbols error when linking on Linux by passing `--code-model=large` to `llc` instead of using `relocation-model=static`.
- `cargo-lib-ci`
  - Fix an ambiguous use when user passes both `--install` and `--uninstall`.
  - Remove checking if the current directory is a valid Cargo directory.
  - Default arguments for the library are changed.

#### Removed

- `cargo` module has been removed. All Cargo related functions have been moved to `ops/build`.
- `wget` usage for fetching the source code has been removed.

## [3.0.1](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/3.0.1)

Released on 2021-07-02.

#### Updated

- Fix the compilation error when running `cargo publish`.
- Version number should now be correct.

## [3.0.0](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/3.0.0)

Released on 2021-07-02.

#### Updated

- `cargo-ci` has been renamed to `cargo-compiler-interrupts`.
- `cargo-build-ci`
  - Temporary fix to the mismatch relocation symbols error when linking on Linux by passing `relocation-model=static` to the compiler.

#### Removed

- Miscellaneous helper functions.

---

## [2.1.0](https://github.com/bitslab/cargo-compiler-interrupts/releases/tag/2.1.0)

Released on 2021-06-14.

#### Added

- `CIError` type for custom errors.
- Helper functions relating to manipulating `PathBuf`.

#### Updated

- Improve checking for incremental compilation.
- Improve error handling when loading the configuration.
- Improve finding the correct LLVM utilities based on LLVM version from `rustc`.
- Use `RUSTC_LOG` to output the internal linker invocation instead of `-Z print-link-args` which requires nightly Rust to work.

#### Removed

- Helper functions relating to running external LLVM utilities.


src/lib.rs
//! [![crates.io](https://img.shields.io/crates/v/cargo-compiler-interrupts.svg)][crates.io]
//! [![docs.rs](https://docs.rs/cargo-compiler-interrupts/badge.svg)][docs.rs]
//! [![license](https://img.shields.io/crates/l/cargo-compiler-interrupts.svg)][license]
//!
//! `cargo-compiler-interrupts` provides you a seamless way to integrate the
//! [Compiler Interrupts][compiler-interrupts-paper] to any Rust packages.
//! Check out the Compiler Interrupts [main repository][compiler-interrupts] for more info.
//!
//! ## Requirements
//!
//! * [Rust 1.45.0 - 1.64.0][rust] and [LLVM 9 - 14][llvm] are required. 
//! Both must have the same LLVM version.
//! Later LLVM versions are currently not supported due to the new LLVM pass manager.
//! * You can check the LLVM version from Rust and LLVM toolchains by running `rustc -vV`
//! and `llvm-config --version` respectively.
//! * x86-64 architecture with Linux or macOS is highly recommended.
//! Other architectures and platforms have not been tested and not guaranteed to work.
//!
//! ## Installation
//!
//! `cargo-compiler-interrupts` can be installed via `cargo install`.
//!
//! ``` sh
//! cargo install cargo-compiler-interrupts
//! ```
//!
//! You can also fetch the repo and install using `--path`.
//!
//! ``` sh
//! git clone https://github.com/bitslab/cargo-compiler-interrupts
//! cargo install --path ./cargo-compiler-interrupts
//! ```
//!
//! ## Getting started
//!
//! `cargo-compiler-interrupts` provides three binaries:
//!
//! ``` sh
//! cargo-lib-ci install    # install the CI library
//! cargo-build-ci          # build and integrate CI to the binary
//! cargo-run-ci            # run the CI-integrated binary
//! ```
//!
//! * `cargo-lib-ci` — manage the Compiler Interrupts library.
//! * `cargo-build-ci` — build and integrate the Compiler Interrupts to the package.
//! * `cargo-run-ci` — run the integrated binary.
//! You can specify which binary to run by passing `--bin <BINARY>`.
//!
//! Run `cargo-lib-ci install` to install the Compiler Interrupts library first.
//! Before running `cargo-build-ci`, add the Compiler Interrupts API package as the dependency for
//! your Cargo package and registers the Compiler Interrupts handler in your program.
//! Compiler Interrupts API is provided through the [`compiler-interrupts`][compiler-interrupts-rs]
//! package.
//!
//! ``` rust
//! fn interrupt_handler(ic: i64) {
//!     println!("Compiler interrupt called with instruction count: {}", ic);
//! }
//!
//! unsafe {
//!     compiler_interrupts::register(1000, 1000, interrupt_handler);
//! }
//! ```
//!
//! For more detailed usages and internals, run the command with `--help` option and
//! check out the **[documentation]**.
//!
//! ## Contribution
//!
//! All issue reports, feature requests, pull requests and GitHub stars are welcomed
//! and much appreciated. Issues relating to the Compiler Interrupts library
//! should be reported to the [main repository][compiler-interrupts].
//!
//! ## Author
//!
//! Quan Tran ([@quanshousio][quanshousio])
//!
//! ## Acknowledgements
//!
//! * My advisor [Jakob Eriksson][jakob] for the enormous support for this project.
//! * [Nilanjana Basu][nilanjana] for implementing the Compiler Interrupts.
//!
//! ## License
//!
//! `cargo-compiler-interrupts` is available under the MIT license.
//! See the [LICENSE][license] file for more info.
//!
//! [crates.io]: https://crates.io/crates/cargo-compiler-interrupts
//! [docs.rs]: https://docs.rs/cargo-compiler-interrupts
//! [license]: https://github.com/bitslab/cargo-compiler-interrupts/blob/main/LICENSE
//! [documentation]: https://github.com/bitslab/cargo-compiler-interrupts/blob/main/DOCUMENTATION.md
//! [compiler-interrupts]: https://github.com/bitslab/CompilerInterrupts
//! [compiler-interrupts-rs]: https://github.com/bitslab/compiler-interrupts-rs
//! [compiler-interrupts-paper]: https://dl.acm.org/doi/10.1145/3453483.3454107
//! [rust]: https://www.rust-lang.org/tools/install
//! [llvm]: https://releases.llvm.org
//! [quanshousio]: https://www.quanshousio.com
//! [jakob]: https://www.linkedin.com/in/erikssonjakob
//! [nilanjana]: https://www.linkedin.com/in/nilanjana-basu-99027959

#![warn(
    missing_copy_implementations,
    missing_debug_implementations,
    missing_docs,
    trivial_casts,
    trivial_numeric_casts,
    unsafe_code,
    unstable_features,
    unused_extern_crates,
    unused_import_braces,
    unused_qualifications,
    clippy::cast_possible_truncation,
    clippy::cast_possible_wrap,
    clippy::cast_precision_loss,
    clippy::cast_sign_loss,
    clippy::clone_on_ref_ptr,
    clippy::missing_docs_in_private_items,
    clippy::mut_mut,
    clippy::print_stdout,
    clippy::unseparated_literal_suffix,
    clippy::unwrap_used
)]

/// Compiler Interrupts result.
type CIResult<T> = anyhow::Result<T>;

mod args;
mod cargo;
mod config;
mod error;
mod llvm;
pub mod ops;
mod paths;
mod util;

/// Name of the cargo-build-ci.
const BUILD_CI_BIN_NAME: &str = "cargo-build-ci";

/// Name of the cargo-run-ci.
const RUN_CI_BIN_NAME: &str = "cargo-run-ci";

/// Name of the cargo-lib-ci.
const LIB_CI_BIN_NAME: &str = "cargo-lib-ci";
